docs/examples/distributed/002_multi_gpu/main.py

```diff
@@ -0,0 +1,271 @@
+
+            optimizer.zero_grad()
+            local_loss.backward()
+            # NOTE: nn.DistributedDataParallel automatically averages the gradients across devices.
+            optimizer.step()
+
+            # Calculate some metrics:
+            # local metrics
+            local_n_correct_predictions = logits.detach().argmax(-1).eq(y).sum()
+            local_n_samples = logits.shape[0]
+            local_accuracy = local_n_correct_predictions / local_n_samples
+
+            # "global" metrics: calculated with the results from all workers
+            # NOTE: Creating new tensors to hold the "global" values, but this isn't required.
+            n_correct_predictions = local_n_correct_predictions.clone()
+            torch.distributed.all_reduce(n_correct_predictions, op=torch.distributed.ReduceOp.SUM)
+            # Actual (global) batch size for this step.
+            n_samples = torch.as_tensor(local_n_samples, device=device)
+            torch.distributed.all_reduce(n_samples, op=torch.distributed.ReduceOp.SUM)
+            # Will store the average loss across all workers.
+            loss = local_loss.clone()
+            torch.distributed.all_reduce(loss, op=torch.distributed.ReduceOp.SUM)
+            loss.div_(world_size)  # Report the average loss across all workers.
+
+            accuracy = n_correct_predictions / n_samples
+
+            logger.debug(f"(local) Accuracy: {local_accuracy:.2%}")
+            logger.debug(f"(local) Loss: {local_loss.item()}")
+            # NOTE: This would log the same values in all workers. Only logging on master:
+            if is_master:
+                logger.debug(f"Accuracy: {accuracy.item():.2%}")
+                logger.debug(f"Average Loss: {loss.item()}")
+
+            # Advance the progress bar one step, and update the "postfix" () the progress bar. (nicer than just)
+            progress_bar.update(1)
+            progress_bar.set_postfix(loss=loss.item(), accuracy=accuracy.item())
+        progress_bar.close()
+
+        val_loss, val_accuracy = validation_loop(model, valid_dataloader, device)
+        # NOTE: This would log the same values in all workers. Only logging on master:
+        if is_master:
+            logger.info(f"Epoch {epoch}: Val loss: {val_loss:.3f} accuracy: {val_accuracy:.2%}")
+
+    print("Done!")
+
+
+@torch.no_grad()
+def validation_loop(model: nn.Module, dataloader: DataLoader, device: torch.device):
+    model.eval()
+
+    total_loss = torch.as_tensor(0.0, device=device)
+    n_samples = torch.as_tensor(0, device=device)
+    correct_predictions = torch.as_tensor(0, device=device)
+
+    for batch in dataloader:
+        batch = tuple(item.to(device) for item in batch)
+        x, y = batch
+
+        logits: Tensor = model(x)
+        loss = F.cross_entropy(logits, y)
+
+        batch_n_samples = x.shape[0]
+        batch_correct_predictions = logits.argmax(-1).eq(y).sum()
+
+        total_loss += loss
```

Comment:
This is for my personal comprehension but why is this line different from the base scrpit?